ace: gpdma: power down routine

tmleman opened the pull request:
We want to power up GPDMA only when we are planning to use it. Device power manager (PM) will wake up the power domain and then power up the DMA instance. When device is no longer in use, PM can disable device and allow the power domain to enter power gaiting state.

tmleman committed:
drivers: gpdma: release dma ownership

Adding function that is allowing to release ownership of the DMA. When
DSP is no longer using dma instance it ownership can be released.

Signed-off-by: Tomasz Leman <tomasz.m.leman@intel.com>

tmleman committed:
drivers: gpdma: enable clock gating

This patch is adding function enabling DMA clock gating.

Signed-off-by: Tomasz Leman <tomasz.m.leman@intel.com>

tmleman committed:
drivers: gpdma: power flow refactor

This patch is refactoring device power flow routine. DMA instance will
not be power up after kernel initialization if device power management
is enabled. Power menager will power up device and its power domain if
device will be in use and disable it when it's no longer needed.

The DMA disabling part has not yet been implemented.

Signed-off-by: Tomasz Leman <tomasz.m.leman@intel.com>

tmleman committed:
drivers: gpdma: power off

This patch allows device power manager to disable GPDMA instance when
its no longer in use.

Signed-off-by: Tomasz Leman <tomasz.m.leman@intel.com>

teburd reviewed (commented):
The changes look good. 

I'd really love to see a test verifying this power logic as part of the change set.

So something like...

1. verify dma is powered off on start
2. verify on start dma is clock enabled and powered on
3. verify on stop dma is clock gated and powered off
4. verify repeated configure/start/stops doesn't break anything.

Perhaps a slight modification of the memory to memory test could do this testing selectively given the particular DMA supports PM and PM is enabled.

softwarecki commented on drivers/power_domain/power_domain_intel_adsp.c:
```diff
@@ -24,6 +24,9 @@ static int pd_intel_adsp_set_power_enable(struct pg_bits *bits, bool power_enabl
 	if (power_enable) {
 		sys_write16(sys_read16((mem_addr_t)&ACE_DfPMCCU.dfpwrctl) | SPA_bit_mask,
 			    (mem_addr_t)&ACE_DfPMCCU.dfpwrctl);
+		while ((sys_read16((mem_addr_t)&ACE_DfPMCCU.dfpwrsts) & BIT(bits->CPA_bit)) == 0) {
```
WAIT_FOR

tmleman replied:
I'm not sure if that's a good solution. There is no scenario where we are waking-up a power domain and it stays off. This would mean a hardware defect, how should we handle it?

abonislawski replied:
This api support errors, device will be not available and application will send error back to host when trying to start gpdma, thats fine for me. 
Anyway not a big deal because of super rare case

marcinszkudlinski commented on drivers/dma/dma_intel_adsp_gpdma.c:
```diff
@@ -219,6 +219,24 @@ static void intel_adsp_gpdma_select_owner(const struct device *dev)
 #endif /* CONFIG_DMA_INTEL_ADSP_GPDMA_NEED_CONTROLLER_OWNERSHIP */
 }
 
+#ifdef CONFIG_PM_DEVICE
```
PM only?

ceolin replied:
yep, that is weird.  Why `intel_adsp_gpdma_claim_ownership` is not unconditionally built while `intel_adsp_gpdma_release_ownership` is only built when `CONFIG_PM_DEVICE` is defined ?

tmleman replied:
If PM_DEVICE is not enabled there is no use of this function.

ceolin commented on drivers/dma/dma_intel_adsp_gpdma.c:
```diff
@@ -317,14 +376,25 @@ int intel_adsp_gpdma_get_attribute(const struct device *dev, uint32_t type, uint
 	case PM_DEVICE_ACTION_TURN_ON:
+		return intel_adsp_gpdma_power_on(dev);
 	case PM_DEVICE_ACTION_TURN_OFF:
+		return intel_adsp_gpdma_power_off(dev);
```
this only happens if this device is under a domain and the domain is suspended

tmleman replied:
I'm aware of it but this looks like the only right place to do this. I want to turn off device and only PD can trigger TURN_OFF action.

tmleman replied:
@ceolin I decide to move power on and off code to SUSPEND/RESUME actions. It may not fit to the name but it's much better to power on only the device that are in use and not all of them, just because they are in the same power domain.

abonislawski reviewed (changes requested):
Blocking just to not accidentally merge without dma_ctx init fix, example:
https://github.com/thesofproject/zephyr/commit/9c22efe46882b10cbf01c67d3af031658a26b703

tmleman committed:
tests: drivers: dma: loop_transfer: power state check

This patch modifies the existing loop transfer test to allow testing for
device power state changes when starting and stopping a transfer.

Signed-off-by: Tomasz Leman <tomasz.m.leman@intel.com>

tmleman committed:
tests: dma: loop_transfer: add mtl board

Adding meteorlake board overlay.

Signed-off-by: Tomasz Leman <tomasz.m.leman@intel.com>

tmleman commented:
> Blocking just to not accidentally merge without dma_ctx init fix

@abonislawski  fixed in newest version

teburd commented:
Thank you so much for adding a test for this, will try it out here today on a few boards

teburd commented:
On cavs25 currently when building with PM, PM_DEVICE, PM_DEVICE_RUNTIME it seems to get stuck on the first memory to memory transfer indefinitely

```
west build -p always -b intel_adsp_cavs25 tests/drivers/dma/loop_transfer -- -DCONFIG_PM=y -DCONFIG_PM_DEVICE=y -DCONFIG_PM_DEVICE_RUNTIME=y
```

Results in

```
*** Booting Zephyr OS build zephyr-v3.3.0-1641-g9583f300912e ***
Running TESTSUITE dma_m2m_loop
===================================================================
START - test_test_dma0_m2m_loop
DMA memory to memory transfer started
Preparing DMA Controller: dma@7c000
Starting the transfer on channel 0 and waiting for 1 second
```

And never making progress, I had to hard restart the board to try again

Without PM enabled this branch doesn't get to the Zephyr boot message.

On main this test works. Please take a look

teburd reviewed (dismissed):
Meant to -1 until tests are working

Thank you for adding them though as it means this feature is easily tested for now.

tmleman commented:
I decided to move the power domains related commits to a separate [pull request](https://github.com/zephyrproject-rtos/zephyr/pull/56233).

softwarecki commented on tests/drivers/dma/loop_transfer/boards/intel_adsp_ace15_mtpm.overlay:
```diff
@@ -0,0 +1,7 @@
+/*
+ * Copyright (c) 2023 Tomasz Leman <tomasz.m.leman@intel.com>
```
Shouldn't the copyright belong to the Intel Corporation and you listed as the author?
@nashif

ceolin commented:
One thing that I think it is wrong with this approach is that you are making the device dependent of PM. PM is supposed to be optional, if you have this option turned off the device should basically be always turned on.

teburd commented:
@tmleman let me know when I should retry/retest, happy to do so

tmleman committed:
drivers: gpdma: pm runtime works only on ace

CAVS platforms are not fully integrated with zephyr. Some of the
registers are still programed from SOF side. This feature can be enabled
for those platforms later when integration is fully done.

Signed-off-by: Tomasz Leman <tomasz.m.leman@intel.com>

tmleman commented:
@teburd unfortunately you won't be able to test it on any CAVS platform. First problem is that builds with enabled PM doesn't work (those without SOF) https://github.com/zephyrproject-rtos/zephyr/issues/56415. And second one is that in case of a TGL some registers are still programed from SOF side.

I made cooperation with closed source version and power management aspects of those drivers does not match to each other. Those features are needed for ACE and I don't want to invest more time into CAVS version.

I will make SOF build with those changes and run test on it.

tmleman commented:
I have tested FW with those changes on MTL and TGL. I didn't find any issue.

https://github.com/thesofproject/sof/pull/7353

teburd commented:
@ceolin please take a look

kv2019i reviewed (approved):
@ceolin Can you check? I'll update https://github.com/zephyrproject-rtos/zephyr/pull/55738 once this is in.

teburd reviewed (approved):
Verified working on cavs25 using the dma loopback test without PM enabled
